quic/s2n-quic-transport/src/dc/manager.rs

```diff
@@ -0,0 +1,242 @@
+    version: Option<dc::Version>,
+    state: State,
+    stateless_reset_token_sync: Flag,
+}
+
+type Flag = flag::Flag<DcStatelessResetTokenWriter>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum State {
+    InitServer,
+    InitClient,
+    ServerPathSecretsReady,
+    ClientPathSecretsReady,
+    ServerTokensSent,
+    Complete,
+}
+
+impl State {
+    is!(is_init, InitServer | InitClient);
+    is!(
+        is_path_secrets_ready,
+        ServerPathSecretsReady | ClientPathSecretsReady
+    );
+    is!(is_server_tokens_sent, ServerTokensSent);
+    is!(is_complete, Complete);
+
+    event! {
+        on_path_secrets_ready(
+            InitServer => ServerPathSecretsReady,
+            InitClient => ClientPathSecretsReady
+        );
+        on_peer_stateless_reset_tokens(
+            ClientPathSecretsReady => Complete,
+            ServerPathSecretsReady => ServerTokensSent
+        );
+        on_stateless_reset_tokens_acked(ServerTokensSent => Complete);
+    }
+}
+
+impl<Config: endpoint::Config> Manager<Config> {
+    /// Constructs a new `dc::Manager` with the optional given path
+    ///
+    /// If path is `None`, the `dc::Manager` will be disabled
+    pub fn new<Pub: event::ConnectionPublisher>(
+        path: Option<<<Config as endpoint::Config>::DcEndpoint as Endpoint>::Path>,
+        version: dc::Version,
+        publisher: &mut Pub,
+    ) -> Self {
+        if path.is_some() {
+            publisher.on_dc_state_changed(DcStateChanged {
+                state: DcState::VersionNegotiated { version },
+            });
+            let state = if Config::ENDPOINT_TYPE.is_server() {
+                State::InitServer
+            } else {
+                State::InitClient
+            };
+
+            Self {
+                path,
+                version: Some(version),
+                state,
+                stateless_reset_token_sync: Flag::default(),
+            }
+        } else {
+            Self::disabled()
+        }
+    }
+
+    /// Returns a disabled `dc::Manager`
+    pub fn disabled() -> Self {
+        Self {
+            path: None,
+            version: None,
+            state: State::Complete,
+            stateless_reset_token_sync: Flag::default(),
+        }
+    }
+
+    /// The dc version that was negotiated, if any
+    ///
+    /// Returns `None` if no version was negotiated or the `dc::Endpoint` did
+    /// not initialize a path for the connection
+    pub fn version(&self) -> Option<dc::Version> {
+        self.version
+    }
+
+    /// Called when the TLS session has indicated path secrets are ready
+    /// to be derived for the dc path
+    ///
+    /// Initiates sending of the `DC_STATELESS_RESET_TOKENS` frame on the client
+    pub fn on_path_secrets_ready<Pub: event::ConnectionPublisher>(
+        &mut self,
+        session: &impl tls::TlsSession,
+        publisher: &mut Pub,
+    ) {
+        if self.state.on_path_secrets_ready().is_ok() {
```

Comment:
Very minor nit: you should be able to unnest these if blocks with `ensure!(state.on_event().is_ok())`

Reply:
ah, good call